File: gachisa-backend/src/main/java/com/gachisa/payment/service/dto/RecoveryPreparation.java
package com.gachisa.payment.service.dto;

import com.gachisa.payment.dto.PaymentResponse;
import com.gachisa.payment.entity.Payment;
import com.gachisa.payment.entity.PaymentAttempt;

public record RecoveryPreparation(
        Long paymentAttemptId,
        String paymentKey,
        boolean queryRequired,
        PaymentResponse existingResponse
) {
    public static RecoveryPreparation query(PaymentAttempt attempt) {
        return new RecoveryPreparation(attempt.getId(), attempt.getPgPaymentKey(), true, null);
    }

    public static RecoveryPreparation skip(Payment payment, PaymentAttempt attempt) {
        return new RecoveryPreparation(
                attempt.getId(), attempt.getPgPaymentKey(), false,
                PaymentResponse.from(payment, attempt));
    }
}

File: gachisa-backend/src/main/kotlin/com/gachisa/payment/service/dto/RecoveryPreparation.kt
package com.gachisa.payment.service.dto

import com.gachisa.payment.dto.PaymentResponse
import com.gachisa.payment.entity.Payment
import com.gachisa.payment.entity.PaymentAttempt

data class RecoveryPreparation(
    val paymentAttemptId: Long,
    val paymentKey: String,
    val queryRequired: Boolean,
    val existingResponse: PaymentResponse?,
) {
    companion object {
        fun query(attempt: PaymentAttempt): RecoveryPreparation = RecoveryPreparation(
            paymentAttemptId = attempt.id!!,
            paymentKey = attempt.pgPaymentKey!!,
            queryRequired = true,
            existingResponse = null,
        )

        fun skip(payment: Payment, attempt: PaymentAttempt): RecoveryPreparation = RecoveryPreparation(
            paymentAttemptId = attempt.id!!,
            paymentKey = attempt.pgPaymentKey!!,
            queryRequired = false,
            existingResponse = PaymentResponse.from(payment, attempt),
        )
    }
}
